File: snake.js
// Initialize game variables
const game_width = 800;
const game_height = 600;
const block_size = 20;
const snake_speed = 15;
const game_duration = 300;  // Game duration in seconds (5 minutes)

// Colors
const BLACK = "#000000";
const WHITE = "#FFFFFF";
const RED = "#FF0000";
const GREEN = "#00FF00";
const BLUE = "#0000FF";

class Snake {
    constructor(color, is_ai = false) {
        this.snake_list = [[
            Math.floor(Math.random() * (game_width - block_size) / block_size) * block_size,
            Math.floor(Math.random() * (game_height - block_size) / block_size) * block_size
        ]];
        this.snake_length = 1;
        this.x = this.snake_list[0][0];
        this.y = this.snake_list[0][1];
        this.x_change = 0;
        this.y_change = 0;
        this.color = color;
        this.is_ai = is_ai;
        this.score = 0;
    }

    update(food) {
        if (this.is_ai) {
            // Implement AI logic here
            let [head_x, head_y] = this.snake_list[this.snake_list.length - 1];
            let food_x = food.x;
            let food_y = food.y;
            
            if (Math.random() < 0.7) {  // 70% chance of polarized movement
                if (Math.abs(head_x - food_x) > Math.abs(head_y - food_y)) {
                    if (head_x < food_x && this.x_change >= 0) {
                        [this.x_change, this.y_change] = [block_size, 0];
                    } else if (head_x > food_x && this.x_change <= 0) {
                        [this.x_change, this.y_change] = [-block_size, 0];
                    } else {
                        this.y_change = head_y < food_y ? block_size : -block_size;
                        this.x_change = 0;
                    }
                } else {
                    if (head_y < food_y && this.y_change >= 0) {
                        [this.x_change, this.y_change] = [0, block_size];
                    } else if (head_y > food_y && this.y_change <= 0) {
                        [this.x_change, this.y_change] = [0, -block_size];
                    } else {
                        this.x_change = head_x < food_x ? block_size : -block_size;
                        this.y_change = 0;
                    }
                }
            } else {  // 30% chance of random movement
                let possible_moves = [[block_size, 0], [-block_size, 0], [0, block_size], [0, -block_size]];
                [this.x_change, this.y_change] = possible_moves[Math.floor(Math.random() * possible_moves.length)];
            }
        }
        
        this.x += this.x_change;
        this.y += this.y_change;

        // Wrap around boundaries
        this.x = (this.x + game_width) % game_width;
        this.y = (this.y + game_height) % game_height;

        this.snake_list.push([this.x, this.y]);
        if (this.snake_list.length > this.snake_length) {
            this.snake_list.shift();
        }
    }

    is_collision() {
        let [head_x, head_y] = this.snake_list[this.snake_list.length - 1];
        return this.snake_list.slice(0, -1).some(block => block[0] === head_x && block[1] === head_y);
    }

    eat_food(food) {
        let [head_x, head_y] = this.snake_list[this.snake_list.length - 1];
        if (head_x === food.x && head_y === food.y) {
            this.snake_length += 1;
            this.score += 1;
            return true;
        }
        return false;
    }

    draw(ctx) {
        this.snake_list.forEach(([x, y]) => {
            ctx.fillStyle = this.color;
            ctx.fillRect(x, y, block_size, block_size);
        });
    }
}

class Food {
    constructor() {
        this.reposition();
    }

    reposition() {
        this.x = Math.round(Math.random() * (game_width - block_size) / block_size) * block_size;
        this.y = Math.round(Math.random() * (game_height - block_size) / block_size) * block_size;
    }

    draw(ctx) {
        ctx.fillStyle = RED;
        ctx.fillRect(this.x, this.y, block_size, block_size);
    }
}

let canvas, ctx, user_snake, ai_snake, food, game_state;

function init_game() {
    canvas = document.getElementById('gameCanvas');
    ctx = canvas.getContext('2d');
    user_snake = new Snake(GREEN);
    ai_snake = new Snake(BLUE, true);
    food = new Food();
    game_state = {
        start_time: Date.now() / 1000,
        is_game_over: false
    };
}

function game_loop() {
    if (!game_state.is_game_over) {
        update_game();
        draw_game();
        requestAnimationFrame(game_loop);
    } else {
        show_game_over();
    }
}

function update_game() {
    let current_time = Date.now() / 1000;
    if (current_time - game_state.start_time < game_duration) {
        user_snake.update(food);
        ai_snake.update(food);

        if (user_snake.is_collision()) {
            user_snake = new Snake(GREEN);
        }
        if (ai_snake.is_collision()) {
            ai_snake = new Snake(BLUE, true);
            ai_snake.score = 0;  // Reset AI score when it collides
        }

        if (user_snake.eat_food(food) || ai_snake.eat_food(food)) {
            food.reposition();
        }
    } else {
        game_state.is_game_over = true;
    }
}

function draw_game() {
    ctx.clearRect(0, 0, game_width, game_height);
    
    user_snake.draw(ctx);
    ai_snake.draw(ctx);
    food.draw(ctx);

    // Display scores and time remaining
    ctx.fillStyle = WHITE;
    ctx.font = '20px Arial';
    ctx.fillText(`User: ${user_snake.score}`, 10, 30);
    ctx.fillText(`AI: ${ai_snake.score}`, game_width - 70, 30);
    let time_remaining = Math.max(0, Math.floor(game_duration - (Date.now() / 1000 - game_state.start_time)));
    ctx.fillText(`Time: ${time_remaining}s`, game_width / 2 - 40, 30);
}

function show_game_over() {
    ctx.fillStyle = BLACK;
    ctx.fillRect(0, 0, game_width, game_height);
    ctx.fillStyle = WHITE;
    ctx.font = '30px Arial';
    ctx.textAlign = 'center';
    if (ai_snake.score > user_snake.score) {
        ctx.fillText("AI will replace you AHAHA - AbdullahAI", game_width / 2, game_height / 2);
    } else {
        ctx.fillText("Game Over!", game_width / 2, game_height / 2);
    }
    ctx.fillText(`Final Score - User: ${user_snake.score}, AI: ${ai_snake.score}`, game_width / 2, game_height / 2 + 40);
    ctx.fillText("Press Space to play again", game_width / 2, game_height / 2 + 80);
}

function handle_key_press(event) {
    if (game_state.is_game_over && event.code === 'Space') {
        init_game();
        game_loop();
    } else {
        if (event.key === 'ArrowLeft' && user_snake.x_change === 0) {
            user_snake.x_change = -block_size;
            user_snake.y_change = 0;
        } else if (event.key === 'ArrowRight' && user_snake.x_change === 0) {
            user_snake.x_change = block_size;
            user_snake.y_change = 0;
        } else if (event.key === 'ArrowUp' && user_snake.y_change === 0) {
            user_snake.y_change = -block_size;
            user_snake.x_change = 0;
        } else if (event.key === 'ArrowDown' && user_snake.y_change === 0) {
            user_snake.y_change = block_size;
            user_snake.x_change = 0;
        }
    }
}

document.addEventListener('keydown', handle_key_press);

window.onload = function() {
    init_game();
    game_loop();
};

File: snake.py
import pygame
import random
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
import os
import time

# Initialize Pygame
pygame.init()

# Screen dimensions
game_width, game_height = 800, 600
width, height = game_width, game_height
screen = pygame.display.set_mode((width, height))
pygame.display.set_caption('User vs AI Snake Game')

# Colors
black = (0, 0, 0)
white = (255, 255, 255)
red = (255, 0, 0)
green = (0, 255, 0)
blue = (0, 0, 255)

# Snake settings
block_size = 20
snake_speed = 15
game_duration = 300  # Game duration in seconds (5 minutes)

# Neural Network Hyperparameters
learning_rate = 0.001
gamma = 0.99  # Discount factor for future rewards
epsilon = 0.1  # Exploration rate

# Font settings
font_style = pygame.font.SysFont(None, 30)

def message(msg, color, y_displace=0):
    mesg = font_style.render(msg, True, color)
    text_rect = mesg.get_rect(center=(width/2, height/2 + y_displace))
    screen.blit(mesg, text_rect)

class NeuralNetwork(nn.Module):
    def __init__(self):
        super(NeuralNetwork, self).__init__()
        self.fc1 = nn.Linear(12, 64)
        self.fc2 = nn.Linear(64, 32)
        self.fc3 = nn.Linear(32, 4)

    def forward(self, x):
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

class Snake:
    def __init__(self, color, is_ai=False):
        self.snake_list = [[random.randint(0, game_width - block_size) // block_size * block_size,
                            random.randint(0, game_height - block_size) // block_size * block_size]]
        self.snake_length = 1
        self.x = self.snake_list[0][0]
        self.y = self.snake_list[0][1]
        self.x_change = 0
        self.y_change = 0
        self.color = color
        self.is_ai = is_ai
        self.score = 0
        if is_ai:
            self.brain = NeuralNetwork()
            self.optimizer = optim.Adam(self.brain.parameters(), lr=learning_rate)
            self.memory = []

    def update(self, food):
        if self.is_ai:
            input_state = self.get_input_state(food)
            action = self.get_action(input_state)
            
            # Implement more random and polarized movement for AI
            head_x, head_y = self.snake_list[-1]
            food_x, food_y = food.x, food.y
            
            # Randomly choose between polarized movement and random movement
            if random.random() < 0.7:  # 70% chance of polarized movement
                if abs(head_x - food_x) > abs(head_y - food_y):
                    if head_x < food_x and self.x_change >= 0:
                        self.x_change, self.y_change = block_size, 0
                    elif head_x > food_x and self.x_change <= 0:
                        self.x_change, self.y_change = -block_size, 0
                    else:
                        self.y_change = block_size if head_y < food_y else -block_size
                        self.x_change = 0
                else:
                    if head_y < food_y and self.y_change >= 0:
                        self.x_change, self.y_change = 0, block_size
                    elif head_y > food_y and self.y_change <= 0:
                        self.x_change, self.y_change = 0, -block_size
                    else:
                        self.x_change = block_size if head_x < food_x else -block_size
                        self.y_change = 0
            else:  # 30% chance of random movement
                possible_moves = [(block_size, 0), (-block_size, 0), (0, block_size), (0, -block_size)]
                self.x_change, self.y_change = random.choice(possible_moves)
            
            # Store the experience for learning
            reward = 0
            if self.eat_food(food):
                reward = 10
            elif self.is_collision():
                reward = -10
            else:
                # Reward moving towards food
                distance_before = abs(head_x - food_x) + abs(head_y - food_y)
                distance_after = abs(self.x + self.x_change - food_x) + abs(self.y + self.y_change - food_y)
                if distance_after < distance_before:
                    reward = 1
                else:
                    reward = -1
            
            next_state = self.get_input_state(food)
            self.memory.append((input_state, action, reward, next_state, not self.is_collision()))
            
            # Learn from experience
            self.learn()
        
        self.x += self.x_change
        self.y += self.y_change

        # Wrap around boundaries
        self.x = self.x % game_width
        self.y = self.y % game_height

        self.snake_list.append([self.x, self.y])
        if len(self.snake_list) > self.snake_length:
            del self.snake_list[0]

    def get_input_state(self, food):
        head_x, head_y = self.snake_list[-1]
        return np.array([
            head_x / game_width,
            head_y / game_height,
            (game_width - head_x) / game_width,
            (game_height - head_y) / game_height,
            (food.x - head_x) / game_width,
            (food.y - head_y) / game_height,
            1 if self.x_change < 0 else 0,
            1 if self.x_change > 0 else 0,
            1 if self.y_change < 0 else 0,
            1 if self.y_change > 0 else 0,
            self.check_collision_ahead(),
            len(self.snake_list) / 100
        ])

    def get_action(self, state):
        if random.random() < epsilon:
            return random.randint(0, 3)
        state = torch.FloatTensor(state).unsqueeze(0)
        with torch.no_grad():
            q_values = self.brain(state)
        return torch.argmax(q_values).item()

    def learn(self):
        if len(self.memory) < 1000:  # Wait until we have enough experiences
            return

        batch = random.sample(self.memory, 32)
        states, actions, rewards, next_states, dones = zip(*batch)

        states = torch.FloatTensor(np.array(states))
        actions = torch.LongTensor(actions)
        rewards = torch.FloatTensor(rewards)
        next_states = torch.FloatTensor(np.array(next_states))
        dones = torch.FloatTensor(dones)

        current_q_values = self.brain(states).gather(1, actions.unsqueeze(1))
        next_q_values = self.brain(next_states).max(1)[0]
        target_q_values = rewards + (gamma * next_q_values * dones)

        loss = F.mse_loss(current_q_values.squeeze(), target_q_values)

        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()

    def check_collision_ahead(self):
        head_x, head_y = self.snake_list[-1]
        next_x = head_x + self.x_change
        next_y = head_y + self.y_change
        for block in self.snake_list[:-1]:
            if block[0] == next_x and block[1] == next_y:
                return 1
        return 0

    def is_collision(self):
        head_x, head_y = self.snake_list[-1]
        for block in self.snake_list[:-1]:
            if block == [head_x, head_y]:
                return True
        return False

    def eat_food(self, food):
        head_x, head_y = self.snake_list[-1]
        if head_x == food.x and head_y == food.y:
            self.snake_length += 1
            self.score += 1
            return True
        return False

    def draw(self):
        for i, block in enumerate(self.snake_list):
            if i == len(self.snake_list) - 1:  # Head
                pygame.draw.rect(screen, self.color, [block[0], block[1], block_size, block_size])
                eye_radius = block_size // 5
                if self.x_change > 0:  # Moving right
                    pygame.draw.circle(screen, white, (block[0] + block_size - eye_radius, block[1] + eye_radius), eye_radius)
                    pygame.draw.circle(screen, white, (block[0] + block_size - eye_radius, block[1] + block_size - eye_radius), eye_radius)
                elif self.x_change < 0:  # Moving left
                    pygame.draw.circle(screen, white, (block[0] + eye_radius, block[1] + eye_radius), eye_radius)
                    pygame.draw.circle(screen, white, (block[0] + eye_radius, block[1] + block_size - eye_radius), eye_radius)
                elif self.y_change < 0:  # Moving up
                    pygame.draw.circle(screen, white, (block[0] + eye_radius, block[1] + eye_radius), eye_radius)
                    pygame.draw.circle(screen, white, (block[0] + block_size - eye_radius, block[1] + eye_radius), eye_radius)
                elif self.y_change > 0:  # Moving down
                    pygame.draw.circle(screen, white, (block[0] + eye_radius, block[1] + block_size - eye_radius), eye_radius)
                    pygame.draw.circle(screen, white, (block[0] + block_size - eye_radius, block[1] + block_size - eye_radius), eye_radius)
            else:
                pygame.draw.rect(screen, self.color, [block[0], block[1], block_size, block_size])

class Food:
    def __init__(self):
        self.reposition()

    def reposition(self):
        self.x = round(random.randrange(0, game_width - block_size) / block_size) * block_size
        self.y = round(random.randrange(0, game_height - block_size) / block_size) * block_size

    def draw(self):
        pygame.draw.rect(screen, red, [self.x, self.y, block_size, block_size])

def game_loop():
    clock = pygame.time.Clock()

    user_snake = Snake(green)
    ai_snake = Snake(blue, is_ai=True)
    food = Food()

    start_time = time.time()

    while time.time() - start_time < game_duration:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                quit()
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_LEFT and user_snake.x_change == 0:
                    user_snake.x_change = -block_size
                    user_snake.y_change = 0
                elif event.key == pygame.K_RIGHT and user_snake.x_change == 0:
                    user_snake.x_change = block_size
                    user_snake.y_change = 0
                elif event.key == pygame.K_UP and user_snake.y_change == 0:
                    user_snake.y_change = -block_size
                    user_snake.x_change = 0
                elif event.key == pygame.K_DOWN and user_snake.y_change == 0:
                    user_snake.y_change = block_size
                    user_snake.x_change = 0

        screen.fill(black)

        user_snake.update(food)
        ai_snake.update(food)

        if user_snake.is_collision():
            user_snake = Snake(green)
        if ai_snake.is_collision():
            ai_snake = Snake(blue, is_ai=True)
            ai_snake.score = 0  # Reset AI score when it collides

        if user_snake.eat_food(food):
            food.reposition()
        elif ai_snake.eat_food(food):
            food.reposition()

        user_snake.draw()
        ai_snake.draw()
        food.draw()

        # Display scores and time remaining
        font = pygame.font.Font(None, 36)
        user_score_text = font.render(f"User: {user_snake.score}", True, white)
        ai_score_text = font.render(f"AI: {ai_snake.score}", True, white)
        time_remaining = int(game_duration - (time.time() - start_time))
        time_text = font.render(f"Time: {time_remaining}s", True, white)
        screen.blit(user_score_text, (10, 10))
        screen.blit(ai_score_text, (game_width - 100, 10))
        screen.blit(time_text, (game_width // 2 - 50, 10))

        pygame.display.update()
        clock.tick(snake_speed)

    if ai_snake.score > user_snake.score:
        screen.fill(black)
        message("AI will replace you AHAHA - AbdullahAI", red)
        message("Play again? (Y/N)", white, 50)
        pygame.display.update()
        
        waiting = True
        while waiting:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    quit()
                if event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_y:
                        return True
                    elif event.key == pygame.K_n:
                        return False
    else:
        message("Game Over!", red)
        pygame.display.update()
        pygame.time.wait(2000)

    return False

# Add a try-except block to handle potential errors
try:
    play_again = True
    while play_again:
        play_again = game_loop()
except Exception as e:
    print(f"An error occurred: {e}")
finally:
    pygame.quit()
    quit()
